src/nnue/layers/affine_transform.h

```diff
@@ -74,113 +74,578 @@ namespace Eval::NNUE::Layers {
+        sum = _mm_add_epi32(sum, _mm_shuffle_epi32(sum, 0xB1)); //_MM_PERM_CDAB
+        return _mm_cvtsi128_si32(sum) + bias;
+      };
+
+      auto m128_haddx4 = [](__m128i sum0, __m128i sum1, __m128i sum2, __m128i sum3, __m128i bias) -> __m128i {
+        sum0 = _mm_hadd_epi32(sum0, sum1);
+        sum2 = _mm_hadd_epi32(sum2, sum3);
+
+        sum0 = _mm_hadd_epi32(sum0, sum2);
+
+        return _mm_add_epi32(sum0, bias);
+      };
+#endif
+
+#if defined (USE_AVX512) && defined (USE_VNNI)
+
       const auto output = reinterpret_cast<OutputType*>(buffer);
 
-  #if defined(USE_AVX512)
+      const auto input_vector256 = reinterpret_cast<const __m256i*>(input);
+      const auto input_vector512 = reinterpret_cast<const __m512i*>(input);
+
       constexpr IndexType kNumChunks = kPaddedInputDimensions / (kSimdWidth * 2);
       const auto input_vector = reinterpret_cast<const __m512i*>(input);
-  #if !defined(USE_VNNI)
-      const __m512i kOnes = _mm512_set1_epi16(1);
-  #endif
 
-  #elif defined(USE_AVX2)
-      constexpr IndexType kNumChunks = kPaddedInputDimensions / kSimdWidth;
-      const auto input_vector = reinterpret_cast<const __m256i*>(input);
-  #if !defined(USE_VNNI)
-      const __m256i kOnes = _mm256_set1_epi16(1);
-  #endif
+      // kOutputDimensions is either 1, 32, or 512
+      if constexpr (kOutputDimensions % 4 == 0)
+      {
+        for (IndexType i = 0; i < kOutputDimensions; i += 4) {
+          const IndexType offset0 = (i + 0) * kPaddedInputDimensions;
+          const IndexType offset1 = (i + 1) * kPaddedInputDimensions;
+          const IndexType offset2 = (i + 2) * kPaddedInputDimensions;
+          const IndexType offset3 = (i + 3) * kPaddedInputDimensions;
+
+          const __m128i bias = _mm_load_si128(reinterpret_cast<const __m128i*>(&biases_[i]));
+          __m128i* outptr = reinterpret_cast<__m128i*>(&output[i]);
+
+          // kPaddedInputDimensions % (kSimdWidth * 2) == 0 is only
+          // false for the last layer. So for the hidden layers
+          // the size is a multiple of 512 bits.
+          if constexpr (kPaddedInputDimensions % (kSimdWidth * 2) == 0)
+          {
+            __m512i sum0 = _mm512_setzero_si512();
+            __m512i sum1 = _mm512_setzero_si512();
+            __m512i sum2 = _mm512_setzero_si512();
+            __m512i sum3 = _mm512_setzero_si512();
+
+            const auto row0 = reinterpret_cast<const __m512i*>(&weights_[offset0]);
+            const auto row1 = reinterpret_cast<const __m512i*>(&weights_[offset1]);
+            const auto row2 = reinterpret_cast<const __m512i*>(&weights_[offset2]);
+            const auto row3 = reinterpret_cast<const __m512i*>(&weights_[offset3]);
+
+            for (IndexType j = 0; j < kNumChunks; ++j) {
+              const __m512i in = _mm512_loadA_si512(&input_vector512[j]);
+
+              sum0 = _mm512_dpbusd_epi32(sum0, in, _mm512_load_si512(&row0[j]));
```

Comment:
I would leave out the explicit load instruction (so just `row0[j]`), and same for aligned stores. It looks cleaner, and it might give the compiler more opportunity for optimisation.